### PR TITLE
added iterative validation calculation

### PR DESCRIPTION
**Pull request by roironn**
_Added iterative validation calculation_

---

_Validation errors are calculated after every nacl files update and cached. The errors are calculated for the changed elements and every element the depends on them or their dependents._

---
_Release Notes_: 
_Replace me with a short sentence that describes the effect of this change on Salto users_

### CODE REVIEW

File: packages/workspace/src/workspace/workspace.ts [hunk 1]
```diff
@@ -371,15 +447,11 @@ export const loadWorkspace = async (
   }
 
   const errors = async (validate = true): Promise<Errors> => {
```
Comment (noamhammer) [inline packages/workspace/src/workspace/workspace.ts:449]: make sure you pass this flag to all the update functions in the workspace

File: packages/lang-server/test/workspace.ts [hunk 2]
```diff
@@ -101,8 +101,8 @@ const persistentMockCreateRemoteMap = ():
       delete: async (key: K) => {
         delete maps[opts.namespace][key]
       },
-      deleteAll: async () => {
-        maps[opts.namespace] = {} as Record<K, string>
+      deleteAll: async (keys: K[]) => {
```
[the remark this answers is not in the document]
Comment (roironn) [reply packages/lang-server/test/workspace.ts:104]: Totally should. We need a function that delete multiple keys :)
Clear is clear and delete is delete :)

[the remark this answers is not in the document]
Comment (noamhammer) [reply packages/lang-server/test/workspace.ts]: maybe rename it to `deleteMany`?

File: packages/workspace/src/workspace/workspace.ts [hunk 3]
```diff
@@ -93,15 +94,15 @@ export type Workspace = {
...
+  errors(): Promise<Readonly<Errors>>
   transformToWorkspaceError<T extends SaltoElementError>(saltoElemErr: T):
     Promise<Readonly<WorkspaceError<T>>>
   transformError: (error: SaltoError) => Promise<WorkspaceError<SaltoError>>
   updateNaclFiles: (changes: DetailedChange[], mode?: RoutingMode) => Promise<void>
   listNaclFiles: () => Promise<string[]>
   getTotalSize: () => Promise<number>
   getNaclFile: (filename: string) => Promise<NaclFile | undefined>
-  setNaclFiles: (...naclFiles: NaclFile[]) => Promise<Change[]>
+  setNaclFiles: (naclFiles: NaclFile[], validate?: boolean) => Promise<Change[]>
```
Comment (noamhammer) [inline packages/workspace/src/workspace/workspace.ts:105]: why you added it only to `setNaclFiles`? what about `removeNaclFiles` for instance?

**Review by noamhammer — APPROVED** [review]

File: packages/workspace/src/workspace/workspace.ts [hunk 4]
```diff
@@ -200,7 +208,59 @@ export const loadWorkspace = async (
...
+          serialize: validationErrors => serialize(validationErrors),
+          deserialize: async data => deserializeValidationErrors(data),
+        }),
+      }
+
+    const getElementsDependents = async (
+      elemIDs: ElemID[],
+      elementSource: ReadOnlyElementsSource,
+      addedIDs: Set<string>
+    ): Promise<Element[]> => {
```
Comment (adamro) [inline packages/workspace/src/workspace/workspace.ts:222]: If this func doesn't return the "base" elemIDs that depend on other, then it probably makes sense to return ElemID[] and not Element[]

File: packages/workspace/test/serializer/serializer.test.ts [hunk 5]
```diff
@@ -428,4 +430,51 @@ describe('State/cache serialization', () => {
       await expect(deserializeMergeErrors(safeJsonStringify([{ test }]))).rejects.toThrow()
     })
   })
+
+  describe('validation errors', () => {
+    const validationErrors = _.sortBy([
+      // new InvalidStaticFileError({
```
Comment (adamro) [inline packages/workspace/test/serializer/serializer.test.ts:436]: ?

File: packages/workspace/test/workspace/workspace.test.ts [hunk 6]
```diff
@@ -1128,8 +1129,8 @@ describe('workspace', () => {
       expect(lead.fields.base_field.annotations[CORE_ANNOTATIONS.DEFAULT]).toEqual('foo')
       expect(lead.fields.new_field).toBeDefined()
     })
-
-    it('should not modify the changes object', () => {
+    // eslint-disable-next-line
```
Comment (adamro) [inline packages/workspace/test/workspace/workspace.test.ts:1132]: why like this?

File: packages/workspace/src/workspace/workspace.ts [hunk 7]
```diff
@@ -288,7 +348,33 @@ export const loadWorkspace = async (
...
+        stateToBuild.merged,
+        changeIDs,
+      )
+      const validationErrorsById = await awu(validationErrors)
+        .groupBy(err => err.elemID.createTopLevelParentID().parent.getFullName())
+
+      const errorsToUpdate = Object.entries(validationErrorsById)
+        .map(([elemID, errors]) => ({ key: elemID, value: errors }))
+
+      const errorsToDelete = validatedElementsIDs
```
Comment (adamro) [inline packages/workspace/src/workspace/workspace.ts:371]: `elementsWithNoErrors`

File: packages/workspace/test/workspace/workspace.test.ts (hunk 6 above)
Comment (roironn) [reply packages/workspace/test/workspace/workspace.test.ts:1132]: removed the test after talking with neta regarding it (she added it for another scenario - change in the elements is now valid (they are being resolved)

**Pull request merged**